File: autres-projets-auxiliaires/binary-flat-file-analyzer/binary-flat-file-analyzer/PMDataScanner.m
//
//  PMDataScanner.m
//  canari
//
//  Created by Pierre Molinaro on 09/11/07.
//  Copyright 2007 __MyCompanyName__. All rights reserved.
//
//---------------------------------------------------------------------------*

#import "PMDataScanner.h"

//#define EASY_BINDINGS_DEBUG

//---------------------------------------------------------------------------*

static NSUInteger uimin (const NSUInteger inA, const NSUInteger inB) {
  return (inA < inB) ? inA : inB ;
}

//---------------------------------------------------------------------------*

@implementation PMDataScanner

//---------------------------------------------------------------------------*

- (void) configureWithData: (NSData *) inData {
    mReadIndex = 0 ;
    mData = inData ;
    mReadOk = YES ;
    mExpectedBytes = [NSMutableData new] ;
}

//---------------------------------------------------------------------------*

+ (PMDataScanner *) scannerWithData: (NSData *) inData {
  PMDataScanner * scanner  = [PMDataScanner new] ;
  [scanner configureWithData:inData] ;
  return scanner ;
}

//---------------------------------------------------------------------------*

- (void) ignoreBytes: (NSUInteger) inLengthToIgnore {
  if (mReadOk) {
    mReadIndex += inLengthToIgnore ;
  }
}

//---------------------------------------------------------------------------*

- (BOOL) testAcceptByte: (UInt8) inByte {
  BOOL result = mReadOk ;
  if (result) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = bytes [mReadIndex] == inByte ;
      if (result) {
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        [mExpectedBytes appendBytes:& inByte length:1] ;
      }
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (BOOL) testAcceptByte: (UInt8) inByte
         withMessage: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  BOOL result = [self testAcceptByte:inByte] ;
  if (result) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    printf (" %02X", inByte) ;
    printf (" | %s\n", [inMessage cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (BOOL) testAcceptFromByte: (UInt8) inLowerBound
         toByte: (UInt8) inUpperBound
         value: (UInt8 *) outValuePtr {
  BOOL result = mReadOk ;
  if (result) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = (bytes [mReadIndex] >= inLowerBound) && (bytes [mReadIndex] <= inUpperBound) ;
      if (result) {
        if (outValuePtr != NULL) {
          * outValuePtr = bytes [mReadIndex] ;
        }
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        for (NSUInteger i=inLowerBound ; i<=inUpperBound ; i++) {
          [mExpectedBytes appendBytes:& i length:1] ;
        }
      }
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (void) acceptRequiredByte: (UInt8) inByte {
  if (mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      if (bytes [mReadIndex] == inByte) {
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        NSMutableString * message = [NSMutableString new] ;
        const UInt8 * expectedBytes = [mExpectedBytes bytes] ;
        for (NSUInteger i=0 ; i<[mExpectedBytes length] ; i++) {
          [message appendFormat:@"0x%02hhx, ", expectedBytes  [i]] ;
        }
        NSLog (@"Invalid current byte (0x%02x): expected bytes:%@0x%02x", bytes [mReadIndex], message, inByte) ;
        mReadOk = NO ;
      }
    }
  }
}

//---------------------------------------------------------------------------*

- (void) acceptRequiredByte: (UInt8) inByte
         withMessage: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  [self acceptRequiredByte:inByte] ;
  if (mReadOk) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    printf (" %02X", inByte) ;
    printf (" | %s\n", [inMessage cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
}

//---------------------------------------------------------------------------*

- (void) acceptRequiredBytes: (UInt8 *) inBytes
         length: (NSUInteger) inLength
         message: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  for (NSUInteger i=0 ; i<inLength ; i++) {
    [self acceptRequiredByte: inBytes [i]] ;
  }
  if (mReadOk) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    const UInt8 * bytes = [mData bytes] ;
    for (NSUInteger i=0 ; i<inLength ; i++) {
      printf (" %02X", bytes [idx + i]) ;
    }
    printf (" | %s\n", [inMessage cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
}

//---------------------------------------------------------------------------*

- (UInt8) parseByte {
  UInt8 result = 0 ;
  if (mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = bytes [mReadIndex] ;
      mReadIndex ++ ;
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (void) parseByteWithMessage: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  const UInt8 byte = self.parseByte ;
  if (mReadOk) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    printf (" %02X", byte) ;
    printf (" | %s\n", [inMessage cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
}

//---------------------------------------------------------------------------*

- (UInt64) parseAutosizedUnsignedInteger {
  UInt64 result = 0 ;
  const UInt8 * bytes = [mData bytes] ;
  NSUInteger shift = 0 ;
  BOOL loop = YES ;
  while (loop && mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
    }else{
      const UInt64 w = bytes [mReadIndex] & 0x7F ;
      result |= (UInt64) (w << shift) ;
      shift += 7 ;
      loop = (bytes [mReadIndex] & 0x80) !=0 ;
      mReadIndex ++ ;
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (NSUInteger) parseAutosizedUnsignedIntegerWithMessage: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  NSUInteger result = [self parseAutosizedUnsignedInteger] ;
  if (result) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    const UInt8 * bytes = [mData bytes] ;
    for (NSUInteger i=idx ; i<mReadIndex ; i++) {
      printf (" %02X", bytes [i]) ;
    }
    NSString * s = [NSString stringWithFormat:@"%@ (size: %lu bytes, value: %lu)", inMessage, mReadIndex - idx, result] ;
    printf (" | %s\n", [s cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (NSString *) parseAutosizedString {
  NSString * result = nil ;
  const UInt8 * bytes = [mData bytes] ;
  NSUInteger beyondStringIndex = mReadIndex ;
  BOOL loop = YES ;
  while (loop && mReadOk) {
    if (beyondStringIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
    }else{
      loop = bytes [beyondStringIndex] != 0 ;
      beyondStringIndex ++ ;
    }
  }
  if (mReadOk) {
    result = [NSString stringWithUTF8String:(const char *) & bytes [mReadIndex]] ;
    mReadIndex = beyondStringIndex ;
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (NSString *) parseAutosizedStringWithMessage: (NSString *) inMessage {
  const NSUInteger idx = mReadIndex ;
  NSString * result = [self parseAutosizedString] ;
  if (result) {
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    const UInt8 * bytes = [mData bytes] ;
    const NSUInteger maxPrint = uimin (idx + 10, mReadIndex) ;
    for (NSUInteger i=idx ; i<maxPrint ; i++) {
      printf (" %02X", bytes [i]) ;
    }
    if (maxPrint < mReadIndex) {
      printf ("...") ;
    }
    NSString * s = [NSString stringWithFormat:@"%@ (size: %lu bytes, value: '%@')", inMessage, mReadIndex - idx, result] ;
    printf (" | %s\n", [s cStringUsingEncoding:NSUTF8StringEncoding]) ;
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (NSData *) parseAutosizedData {
  NSData * result = nil ;
  if (mReadOk) {
    const NSUInteger dataLength = [self parseAutosizedUnsignedInteger] ;
    if ((mReadIndex + dataLength) >= [mData length]) {
      NSLog (@"Read beyond end of data") ;
      mReadOk = NO ;
    }else{
      result = [mData subdataWithRange:NSMakeRange (mReadIndex, dataLength)] ;
      mReadIndex += dataLength ;
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (NSData *) parseAutosizedDataWithMessage: (NSString *) inMessage {
  NSData * result = nil ;
  if (mReadOk) {
    const NSUInteger idx = mReadIndex ;
    const NSUInteger dataLength = [self parseAutosizedUnsignedInteger] ;
    printf ("%04lX %04lX:", (idx >> 32), idx % (1UL << 32)) ;
    const UInt8 * bytes = [mData bytes] ;
    for (NSUInteger i=idx ; i<mReadIndex ; i++) {
      printf (" %02X", bytes [i]) ;
    }
    NSString * s = [NSString stringWithFormat:@"%@ (length field, size: %lu bytes)", inMessage, mReadIndex - idx] ;
    printf (" | %s\n", [s cStringUsingEncoding:NSUTF8StringEncoding]) ;
    if ((mReadIndex + dataLength) >= [mData length]) {
      NSLog (@"Read beyond end of data") ;
      mReadOk = NO ;
    }else{
      const NSUInteger idx2 = mReadIndex ;
      result = [mData subdataWithRange:NSMakeRange (mReadIndex, dataLength)] ;
      mReadIndex += dataLength ;
      printf ("%04lX %04lX:", (idx2 >> 32), idx2 % (1UL << 32)) ;
      s = [NSString stringWithFormat:@"%@ (contents, size: %lu bytes)", inMessage, dataLength] ;
      printf ("... | %s\n", [s cStringUsingEncoding:NSUTF8StringEncoding]) ;
    }
  }
  return result ;
}

//---------------------------------------------------------------------------*

- (void) ignoreAutosizedData {
  if (mReadOk) {
    const NSUInteger dataLength = [self parseAutosizedUnsignedInteger] ;
    [self ignoreBytes:dataLength] ;
  }
}

//---------------------------------------------------------------------------*

- (BOOL) ok {
  return mReadOk ;
}

//---------------------------------------------------------------------------*

@end

File: generation-templates/swift-sources/PMDataScanner.swift
//
//  PMDataScanner.m
//  canari
//
//  Created by Pierre Molinaro on 09/11/07.
//  Copyright 2007 __MyCompanyName__. All rights reserved.
//
//---------------------------------------------------------------------------*

import Cocoa

//---------------------------------------------------------------------------*

@objc(PMDataScanner) class PMDataScanner : NSObject {
  var mData : NSData
  var mReadIndex : Int = 0
  var mReadOk : Bool = true
  var mExpectedBytes : Array<UInt8> = []
  var mProgressWindow : NSWindow? = nil
  var mProgressIndicator : NSProgressIndicator? = nil

  //---------------------------------------------------------------------------*
  //  init                                                                     *
  //---------------------------------------------------------------------------*

  init (data: NSData,
        displayProgressWindowTitle: String?) {
    mData = data
    super.init ()
    noteObjectAllocation (self)
    if (nil != displayProgressWindowTitle) {
      openProgressWindowWithTitle (displayProgressWindowTitle!)
    }
  }

  //---------------------------------------------------------------------------*
  //  deinit                                                                   *
  //---------------------------------------------------------------------------*

  deinit {
    mProgressWindow?.orderOut (self)
    noteObjectDeallocation (self)
  }

  //---------------------------------------------------------------------------*
  //  openProgressWindowWithTitle                                              *
  //---------------------------------------------------------------------------*

  func openProgressWindowWithTitle (inTitle : String) {
    let visibleFrame = NSScreen.mainScreen ().visibleFrame
    let windowWidth = 400.0
    let windowHeight = 65.0
    let windowRect = NSMakeRect (
      NSMidX (visibleFrame) - CGFloat (windowWidth / 2.0),
      NSMidY (visibleFrame) - CGFloat (windowHeight / 2.0),
      CGFloat (windowWidth),
      CGFloat (windowHeight)
    )
    mProgressWindow = NSWindow (
      contentRect:windowRect,
      styleMask:NSTitledWindowMask,
      backing:NSBackingStoreType.Buffered,
      defer:false
    )
    mProgressWindow!.excludedFromWindowsMenu = true
    mProgressWindow!.title = "Progress"
    let contientViewRect : NSRect = mProgressWindow!.contentView.frame
  //--- Add comment text
    let ts_r = NSRect (
      x:25.0,
      y:30.0,
      width:NSMaxX (contientViewRect) - 40.0,
      height:20.0
    )
    var ts = NSTextField (frame:ts_r)
    ts.font = NSFont.boldSystemFontOfSize (NSFont.smallSystemFontSize())
    ts.stringValue = NSString (format:"Opening %@…", inTitle)
    ts.bezeled = false
    ts.bordered = false
    ts.editable = false
    ts.drawsBackground = false
    mProgressWindow!.contentView.addSubview (ts)
  //--- Add progress indicator
    let ps_r = NSRect (
      x:20.0,
      y:10.0,
      width:NSMaxX (contientViewRect) - 40.0,
      height: 20.0
    )
    mProgressIndicator = NSProgressIndicator (frame:ps_r)
    mProgressIndicator!.indeterminate = true
    mProgressWindow!.contentView.addSubview (mProgressIndicator!)
  //---
    mProgressIndicator!.minValue = 0.0
    mProgressIndicator!.maxValue = Double (mData.length)
    mProgressIndicator!.doubleValue = Double (mReadIndex)
    mProgressIndicator!.indeterminate = false
    mProgressIndicator!.display ()
  //---
    mProgressWindow!.makeKeyAndOrderFront (nil)
  }

  //---------------------------------------------------------------------------*
  //  updateProgressIndicator                                                  *
  //---------------------------------------------------------------------------*

  func updateProgressIndicator () {
    if let progressIndicator = mProgressIndicator {
      progressIndicator.doubleValue = Double (mReadIndex)
      progressIndicator.display ()
    }
  }

  //---------------------------------------------------------------------------*
  //  ignoreBytes                                                              *
  //---------------------------------------------------------------------------*

  func ignoreBytes (inLengthToIgnore : Int) {
    if mReadOk {
      mReadIndex += inLengthToIgnore ;
      updateProgressIndicator ()
    }
  }

  //---------------------------------------------------------------------------*
  //  testAcceptByte                                                           *
  //---------------------------------------------------------------------------*
  // http://stackoverflow.com/questions/24067085/pointers-pointer-arithmetic-and-raw-data-in-swift

  func testAcceptByte (inByte : UInt8) -> Bool {
    var result = mReadOk
    if result {
      if mReadIndex >= mData.length {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData).memory
     //   let ptr = offsetPointer (mData.bytes, CInt(mReadIndex))
     //   var array : COpaquePointer = mData.bytes ()
     //   var byteArray : CConstPointer<UInt8> = CConstPointer<UInt8> (mData.bytes ())
        result = byte == inByte
        if result {
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          mExpectedBytes.append (inByte)
        }
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  testAcceptFromByte                                                       *
  //---------------------------------------------------------------------------*

  func testAcceptFromByte (lowerBound: UInt8,
                           upperBound: UInt8,
                           inout value:UInt8) -> Bool {
    var result = mReadOk
    if result {
      if mReadIndex >= mData.length {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData).memory
        result = (byte >= lowerBound) && (byte <= upperBound) ;
        if (result) {
          value = byte
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          for i in lowerBound ..< upperBound + 1 {
            mExpectedBytes.append (i)
          }
        }
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  acceptRequiredByte                                                       *
  //---------------------------------------------------------------------------*

  func acceptRequiredByte (inByte : UInt8,
                           sourceFile: String) {
    if mReadOk {
      if mReadIndex >= mData.length {
         NSLog ("Read beyond end of data")
         mReadOk = false
      }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData).memory
        if (byte == inByte) {
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          var message = ""
          for b in mExpectedBytes {
            message += NSString (format:"0x%02hhx, ", b)
          }
          NSLog ("%s: invalid current byte (0x%02x): expected bytes:%@0x%02x", sourceFile, byte, message, inByte) ;
          mReadOk = false
        }
      }
    }
    updateProgressIndicator ()
  }

  //---------------------------------------------------------------------------*
  //  parseByte                                                                *
  //---------------------------------------------------------------------------*

  func parseByte () -> UInt8 {
    var result : UInt8 = 0
    if mReadOk {
      if mReadIndex >= mData.length {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        result = UnsafePointer<UInt8> (byteAsData).memory
        mReadIndex += 1
      }
    }
    updateProgressIndicator ()
    return result
  }

  //---------------------------------------------------------------------------*
  //  parseAutosizedUnsignedInteger                                            *
  //---------------------------------------------------------------------------*

  func parseAutosizedUnsignedInteger () -> UInt {
    var result : UInt = 0
    var shift : UInt = 0
    var loop = true
    while loop && mReadOk {
      if mReadIndex >= mData.length {
         NSLog ("Read beyond end of data")
         mReadOk = false
      }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData).memory
        let w : UInt = UInt (byte) & 0x7F
        result |= (w << shift)
        shift += 7
        loop = (byte & 0x80) != 0
        mReadIndex += 1
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  parseAutosizedData                                                       *
  //---------------------------------------------------------------------------*

  func parseAutosizedData () -> NSData {
  var result = NSData ()
  if mReadOk {
    let dataLength : Int = Int (parseAutosizedUnsignedInteger ())
    if (mReadIndex + dataLength) >= mData.length {
      NSLog ("Read beyond end of data")
      mReadOk = false
    }else{
      result = mData.subdataWithRange (NSMakeRange (mReadIndex, dataLength))
      mReadIndex += dataLength
    }
  }
  updateProgressIndicator ()
  return result ;
}

  //---------------------------------------------------------------------------*
  //  ok                                                                       *
  //---------------------------------------------------------------------------*

  func ok () -> Bool {
    return mReadOk
  }

}

//---------------------------------------------------------------------------*